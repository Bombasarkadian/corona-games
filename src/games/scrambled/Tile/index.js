import React from "react";
import PropTypes from "prop-types";
import { BONUSES } from "../config";

import styles from "./index.module.scss";
import clsx from "clsx";

const propTypes = {
  letter: PropTypes.string,
  replacement: PropTypes.string,
  points: PropTypes.number,
  raised: PropTypes.bool,
  separate: PropTypes.bool,
  highlighted: PropTypes.bool,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  bonus: PropTypes.object,
};
const Tile = ({
  letter,
  replacement,
  points,
  raised,
  separate,
  highlighted,
  disabled,
  bonus,
  onClick,
}) => (
  <div
    onClick={onClick}
    className={clsx({
      [styles.tile]: true,
      [styles.tile_separate]: separate,
      [styles.tile_raised]: raised,
      [styles.tile_highlighted]: highlighted,
      [styles.tile_clickable]: onClick || highlighted,
      [styles.tile_disabled]: disabled,
    })}
  >
    {letter ? (
      <div className={styles.letter}>{letter}</div>
    ) : (
      <div className={clsx(styles.letter, styles.letter_replacement)}>{replacement}</div>
    )}
    {bonus && (
      <div
        className={styles.bonus}
        style={{
          background: BONUSES[bonus.type][bonus.multiply],
        }}
      ></div>
    )}
    {bonus && bonus.type === "letter" ? (
      <div
        className={styles.points}
        style={{
          WebkitTextStroke: `2px ${BONUSES[bonus.type][bonus.multiply]}`,
        }}
      >
        {points * bonus.multiply}
      </div>
    ) : (
      <div className={styles.points}>{points}</div>
    )}
  </div>
);
Tile.propTypes = propTypes;

export default Tile;
