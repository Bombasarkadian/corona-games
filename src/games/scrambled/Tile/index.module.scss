.tile {
  cursor: auto;
  opacity: 1;
  user-select: none;
  border-radius: 5px;
  box-shadow: 3px 3px 0 0 #393434, 2px 2px 0 0 #393434, 1px 1px 0 0 #393434;
  background-color: #655d5d;
  color: #ffffff;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 46px;
  height: 46px;
  position: relative;
  top: -3px;
  left: -3px;
  font-weight: 600;
  margin: 0;
  transition: top 0.3s, box-shadow 0.3s, opacity 0.3s, background-color 0.3s;

  &_raised {
    box-shadow: 3px 3px 0 0 #393434, 2px 2px 0 0 #393434, 1px 1px 0 0 #393434,
      rgba(0, 0, 0, 0.25) 3px 8px 4px 1px;
    background-color: #8f8880;
    top: -8px;
  }

  &_highlighted {
    background-color: #8f8880;
  }

  &_separate {
    margin: 0 5px 0 0;
  }

  &_clickable {
    cursor: pointer;
  }

  &_disabled {
    opacity: 0.5;
  }

  .letter {
    font-size: 22px;

    &_replacement {
      opacity: 0.5;
    }
  }

  .bonus {
    position: absolute;
    bottom: 0;
    left: 0;
    margin: 3px;
    border-radius: 3px;
    width: 6px;
    height: 6px;
  }

  .points {
    position: absolute;
    right: 0;
    bottom: 0;
    margin: 2px;
    font-size: 13px;
    line-height: 1;
    -webkit-text-fill-color: #ffffff;
    paint-order: stroke fill;
  }
}
