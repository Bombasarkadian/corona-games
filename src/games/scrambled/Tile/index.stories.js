import React from "react";
import { text, number, boolean, select } from "@storybook/addon-knobs";

import Tile from ".";
import { action } from "@storybook/addon-actions";

export default {
  component: Tile,
  title: "Scrambled/Tile",
  excludeStories: /.*Data$/,
};

export const Default = () => (
  <Tile
    letter={text("Letter", "L")}
    replacement={text("Replacement", "")}
    points={number("Points", 5)}
    raised={boolean("Raised?", false)}
    highlighted={boolean("Highlighted?", false)}
    disabled={boolean("Disabled?", false)}
    onClick={action("onClick")}
    preview={boolean("Preview", false)}
    bonus={select("Bonus", {
      Nothing: null,
      "Letter x3": { type: "letter", multiply: 3 },
      "Letter x2": { type: "letter", multiply: 2 },
      "Word x3": { type: "word", multiply: 3 },
      "Word x2": { type: "word", multiply: 2 },
    })}
  />
);
