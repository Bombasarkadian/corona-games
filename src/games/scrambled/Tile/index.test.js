import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import Tile from "./index";

describe("scrambled Tile component", () => {
  it("renders correctly", () => {
    expect.hasAssertions();
    const component = renderer.create(
      <Tile
        letter={"X"}
        points={5}
        bonus={{ type: "letter", multiply: 3 }}
        raised
        highlighted
        separate
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with a letter", () => {
    expect.hasAssertions();
    render(<Tile letter={"letter-X"} />);
    expect(screen.queryByText("letter-X")).toBeInTheDocument();
  });

  it("renders with points", () => {
    expect.hasAssertions();
    render(<Tile points={15} />);
    expect(screen.getByText("15")).toBeInTheDocument();
  });

  it("renders with a replacement", () => {
    expect.hasAssertions();
    render(<Tile replacement={"replacement-Y"} />);
    expect(screen.getByText("replacement-Y")).toBeInTheDocument();
  });
});
