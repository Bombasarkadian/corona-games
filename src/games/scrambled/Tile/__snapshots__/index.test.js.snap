// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrambled Tile component renders correctly 1`] = `
<div
  className="tile tile_separate tile_raised tile_highlighted tile_clickable"
>
  <div
    className="letter"
  >
    X
  </div>
  <div
    className="bonus"
    style={
      Object {
        "background": "#16919F",
      }
    }
  />
  <div
    className="points"
    style={
      Object {
        "WebkitTextStroke": "2px #16919F",
      }
    }
  >
    15
  </div>
</div>
`;
